File: FREQNESS_Toolbox/FREQNESS_Startup.m
function [allData, MNI, path_home] = FREQNESS_Startup()
% ========================================================================
%  FREQNESS STARTUP FUNCTION
% ========================================================================
%
%  If you use this toolbox, please cite:
%  Rosso, M., Fernández‐Rubio, G., Keller, P. E., Brattico, E., Vuust, P.,
%  Kringelbach, M. L., & Bonetti, L. (2025).
%  FREQ-NESS Reveals the Dynamic Reconfiguration of Frequency-Resolved
%  Brain Networks During Auditory Stimulation.
%  Advanced Science, 2413195.
%  https://doi.org/10.1002/advs.202413195
%
% ========================================================================
%
%  This function initializes the FREQNESS toolbox by automatically
%  determining its installation directory and adding all required
%  subfolders to the MATLAB path.
%
%  AUTHORS:
%  Mattia Rosso & Leonardo Bonetti
%  mattia.rosso@clin.au.dk
%  leonardo.bonetti@clin.au.dk; leonardo.bonetti@psych.ox.ac.uk
%  Center for Music in the Brain, Aarhus University
%  Centre for Eudaimonia and Human Flourishing, Linacre College, University of Oxford
%
% ========================================================================

%% Define main path based on the directory of this function
path_home = fileparts(mfilename('fullpath'));

%% Add toolbox directories
addpath(path_home);
addpath(fullfile(path_home, 'FREQNESS_Functions'));
addpath(fullfile(path_home, 'FREQNESS_ExternalFunctions'));
addpath(fullfile(path_home, 'FREQNESS_ExternalFunctions', 'nifti_tools'));
addpath(fullfile(path_home, 'FREQNESS_Data'));
addpath(fullfile(path_home, 'FREQNESS_MNI_Coordinates'));

fprintf('\nFREQNESS Startup: loading data and MNI coordinates.\n');
fprintf('\nFREQNESS successfully initialized.\n');
fprintf('Base directory: %s\n\n', path_home);

%% Fetch data from experimental Groups or Conditions

% Find folders inside FREQNESS_Data
folder_data = dir(fullfile(path_home, 'FREQNESS_Data'));
folder_data = folder_data([folder_data.isdir] & ~ismember({folder_data.name},{'.','..'}));

% Prepare container for group / condition data
nfolds  = numel(folder_data);
allData = cell(nfolds,1);
nsubs   = zeros(nfolds,1);   % number of subjects per folder

for foldi = 1:nfolds

    % Current group/condition folder
    this_folder = fullfile(folder_data(foldi).folder, folder_data(foldi).name);

    % List .mat files (one per participant)
    files_mat = dir(fullfile(this_folder, '*.mat'));
    nsubs(foldi) = numel(files_mat);
    % Check whether the files are missing
    if nsubs(foldi) == 0
        warning('No .mat files found in folder: %s', this_folder);
        allData{foldi} = [];
    else

        % Store all participants in this folder
        for subi = 1:nsubs(foldi)

            % Load .mat file (must contain exactly one data matrix)
            S  = load(fullfile(files_mat(subi).folder, files_mat(subi).name));
            fn = fieldnames(S);
            if numel(fn) ~= 1
                error('File %s must contain exactly one data matrix.', files_mat(subi).name);
            end
            this_data = S.(fn{1});   % assign voxels-by-time data matrix

            % On first subject, initialize 3D array
            if subi == 1
                [nvoxs, ntime] = size(this_data);
                groupData = nan(nvoxs, ntime, nsubs(foldi));
            else
                % Basic consistency check
                if ~isequal(size(this_data), [nvoxs ntime])
                    error('Size mismatch in file %s relative to first subject in folder %s.', ...
                        files_mat(subi).name, this_folder);
                end
            end

            % Assign into group matrix along 3rd dimension
            groupData(:,:,subi) = this_data;
        end

        % Store in cell array (one entry per group / condition)
        allData{foldi} = groupData;

    end

end



%% Fetch MNI coordinates (if available)

% Initialize default: empty
MNI = [];

% Find .mat file containing the MNI coordinates
folder_mni = fullfile(path_home, 'FREQNESS_MNI_Coordinates');
files_mni  = dir(fullfile(folder_mni, '*.mat'));

% Proceed only if exactly one file is found
if numel(files_mni) == 1

    S  = load(fullfile(files_mni(1).folder, files_mni(1).name));
    fn = fieldnames(S);

    % Must contain exactly one variable
    if numel(fn) == 1
        temp_mni = S.(fn{1});

        % Must contain a real numeric XYZ coordinate matrix
        if ~isnumeric(temp_mni) || ~isreal(temp_mni) || isempty(temp_mni) || ...
                any(isinf(temp_mni(:)))
            warning(['MNI file %s ignored: coordinates must be provided as ' ...
                'a real numeric matrix without infinite values.'],files_mni(1).name);
        elseif size(temp_mni,2) == 3
            MNI = temp_mni;
        elseif size(temp_mni,1) == 3 && size(temp_mni,2) ~= 3
            MNI = temp_mni'; % transpose if has coordinates along the rows
        else
            warning('MNI file %s ignored: coordinate matrix does not have 3 columns.', ...
                files_mni(1).name);
        end
    else
        warning('MNI file %s ignored: file must contain exactly one variable, consisting of a 3-column coordinate matrix.', ...
            files_mni(1).name);
    end

elseif numel(files_mni) > 1
    warning('Multiple MNI .mat files found. None loaded.');

end

% Ensure that the coordinate grid matches every non-empty dataset
if ~isempty(MNI)
    data_nvoxs = [];
    for foldi = 1:nfolds
        if ~isempty(allData{foldi})
            data_nvoxs(end+1) = size(allData{foldi},1); %#ok<AGROW>
        end
    end

    if ~isempty(data_nvoxs) && any(data_nvoxs ~= size(MNI,1))
        warning(['MNI coordinates ignored: the coordinate file contains %d rows, ' ...
            'but the loaded datasets contain %s voxels. Replace the coordinate ' ...
            'file with one matching the voxel order and source space of your data.'], ...
            size(MNI,1),mat2str(unique(data_nvoxs)));
        MNI = [];
    end
end


end

File: FREQNESS_Toolbox/python/src/freqness/FREQNESS_Startup.py
"""Dataset and MNI-coordinate discovery for FREQ-NESS.

If you use this toolbox, please cite:
Rosso, M., Fernández-Rubio, G., Keller, P. E., Brattico, E., Vuust, P.,
Kringelbach, M. L., & Bonetti, L. (2025). FREQ-NESS Reveals the Dynamic
Reconfiguration of Frequency-Resolved Brain Networks During Auditory
Stimulation. Advanced Science, 2413195.
https://doi.org/10.1002/advs.202413195

Toolbox authors
---------------
Mattia Rosso - Center for Music in the Brain, Aarhus University -
mattia.rosso@clin.au.dk
Chiara Malvaso - University of Bologna - chiara.malvaso2@unibo.it
Leonardo Bonetti - Center for Music in the Brain, Aarhus University; Centre
for Eudaimonia and Human Flourishing, Linacre College, University of Oxford -
leonardo.bonetti@clin.au.dk; leonardo.bonetti@psych.ox.ac.uk
"""

from __future__ import annotations

from pathlib import Path
from typing import TypeAlias
import warnings

import numpy as np
from numpy.typing import NDArray
from scipy.io import loadmat


NumericArray: TypeAlias = NDArray[np.number]


def _mat_variables(path: Path) -> dict[str, NumericArray]:
    """Load user variables from a non-v7.3 MATLAB file."""
    try:
        contents = loadmat(path)
    except NotImplementedError as exc:
        raise ValueError(
            f"{path.name} uses MATLAB v7.3/HDF5 format, which is not yet "
            "supported by FREQNESS_Startup. Save it as MATLAB v7 or earlier."
        ) from exc

    return {
        name: value
        for name, value in contents.items()
        if not name.startswith("__")
    }


def _resolve_toolbox_root(path_home: str | Path | None) -> Path:
    if path_home is not None:
        root = Path(path_home).expanduser().resolve()
        if not root.is_dir():
            raise FileNotFoundError(f"FREQ-NESS toolbox directory not found: {root}")
        return root

    candidates = [Path.cwd(), *Path(__file__).resolve().parents]
    for candidate in candidates:
        if (candidate / "FREQNESS_Data").is_dir() and (
            candidate / "FREQNESS_MNI_Coordinates"
        ).is_dir():
            return candidate

    raise FileNotFoundError(
        "Could not locate FREQNESS_Data and FREQNESS_MNI_Coordinates. "
        "Pass the FREQNESS_Toolbox directory to FREQNESS_Startup(path_home)."
    )


def _load_groups(data_directory: Path) -> list[NumericArray | None]:
    if not data_directory.is_dir():
        raise FileNotFoundError(f"FREQ-NESS data directory not found: {data_directory}")

    groups: list[NumericArray | None] = []
    group_directories = sorted(path for path in data_directory.iterdir() if path.is_dir())

    if not group_directories:
        warnings.warn(
            f"No dataset folders found in {data_directory}. Add one folder per "
            "condition containing participant .mat files.",
            UserWarning,
            stacklevel=2,
        )

    for group_directory in group_directories:
        participant_files = sorted(group_directory.glob("*.mat"))
        if not participant_files:
            warnings.warn(
                f"No .mat files found in folder: {group_directory}",
                UserWarning,
                stacklevel=2,
            )
            groups.append(None)
            continue

        participants: list[NumericArray] = []
        reference_shape: tuple[int, int] | None = None

        for participant_file in participant_files:
            variables = _mat_variables(participant_file)
            if len(variables) != 1:
                raise ValueError(
                    f"File {participant_file.name} must contain exactly one data matrix."
                )

            participant = np.asarray(next(iter(variables.values())))
            if participant.ndim != 2 or not np.issubdtype(participant.dtype, np.number):
                raise ValueError(
                    f"File {participant_file.name} must contain one numeric "
                    "variables-by-time matrix."
                )

            if reference_shape is None:
                reference_shape = participant.shape
            elif participant.shape != reference_shape:
                raise ValueError(
                    f"Size mismatch in file {participant_file.name} relative to "
                    f"the first participant in {group_directory}."
                )

            participants.append(participant)

        groups.append(np.stack(participants, axis=2))

    return groups


def _load_mni_coordinates(mni_directory: Path) -> NumericArray | None:
    if not mni_directory.is_dir():
        raise FileNotFoundError(
            f"FREQ-NESS MNI-coordinate directory not found: {mni_directory}"
        )

    mni_files = sorted(mni_directory.glob("*.mat"))
    if not mni_files:
        return None
    if len(mni_files) > 1:
        warnings.warn(
            "Multiple MNI .mat files found. None loaded.",
            UserWarning,
            stacklevel=2,
        )
        return None

    variables = _mat_variables(mni_files[0])
    if len(variables) != 1:
        warnings.warn(
            f"MNI file {mni_files[0].name} ignored: the file must contain "
            "exactly one coordinate matrix.",
            UserWarning,
            stacklevel=2,
        )
        return None

    coordinates = np.asarray(next(iter(variables.values())))
    if coordinates.ndim != 2 or not np.issubdtype(coordinates.dtype, np.number):
        warnings.warn(
            f"MNI file {mni_files[0].name} ignored: coordinates must be numeric.",
            UserWarning,
            stacklevel=2,
        )
        return None

    if coordinates.shape[1] == 3:
        return coordinates
    if coordinates.shape[0] == 3 and coordinates.shape[1] != 3:
        return coordinates.T

    warnings.warn(
        f"MNI file {mni_files[0].name} ignored: coordinate matrix does not "
        "have three columns.",
        UserWarning,
        stacklevel=2,
    )
    return None


def FREQNESS_Startup(
    path_home: str | Path | None = None,
) -> tuple[list[NumericArray | None], NumericArray | None, Path]:
    """Load convention-based FREQ-NESS datasets and optional MNI coordinates.

    Parameters
    ----------
    path_home
        Path to ``FREQNESS_Toolbox``. When omitted, the function searches the
        current directory and the installed module's parent directories.

    Returns
    -------
    allData
        One entry per group or condition. Non-empty entries have shape
        ``(variables, time, participants)``. Empty group folders are ``None``.
    MNI
        MNI coordinates with shape ``(voxels, 3)``, or ``None`` when no valid
        coordinate file is available.
    path_home
        Resolved toolbox directory as a :class:`pathlib.Path`.

    Notes
    -----
    Unlike MATLAB startup, this function does not modify import paths. Python
    package installation provides that functionality without runtime side
    effects.
    """
    root = _resolve_toolbox_root(path_home)
    all_data = _load_groups(root / "FREQNESS_Data")
    mni = _load_mni_coordinates(root / "FREQNESS_MNI_Coordinates")
    return all_data, mni, root


__all__ = ["FREQNESS_Startup"]
